### PR TITLE
Change signature validation logic to call first without the file content

### PR DESCRIPTION
**Pull request by JamesWTruher**
Only validate the signature with the content if the first non-content validation fails

# PR Summary

When ExecutionPolicy is set to require a signature, under some circumstances a signature placed in a file will not execute with a "hash mismatch" error, even though `Get-AuthenticodeSignature` will return that the signature is valid.
This is because `Get-AuthenticodeSignature` and our invocation path are not symmetrical. This PR adds the same check that we do in `Get-AuthenticodeSignature` into our invocation path.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 1]
```diff
@@ -516,11 +516,23 @@ private static void UntrustPublisher(Signature signature)
             }
         }
 
+        // Check the signature via the SIP which should never erroneously validate an invalid signature
+        // or altered script.
         private static Signature GetSignatureWithEncodingRetry(string path, ExternalScriptInfo script)
         {
+            // Invoke the SIP directly with the most simple method
+            Signature signature = SignatureHelper.GetSignature(path, null);
```
Comment (iSazonov) [inline src/System.Management.Automation/security/SecurityManager.cs:524]:
```suggestion
            Signature signature = SignatureHelper.GetSignature(path, fileContent: null);
```

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 2]
```diff
@@ -516,11 +516,23 @@ private static void UntrustPublisher(Signature signature)
...
+            if (signature.Status == SignatureStatus.Valid)
+            {
+                return signature;
+            }
+
+            // try harder to validate the signature by being explicit about encoding
+            // and providing the script contents
             string verificationContents = System.Text.Encoding.Unicode.GetString(script.OriginalEncoding.GetPreamble()) + script.ScriptContents;
```
Comment (rjmholt) [inline src/System.Management.Automation/security/SecurityManager.cs:532]: Can we add a `using System.Text;` to the top for this?

**Review by TravisEz13 — APPROVED** [review]
I believe this is correct.  
`GetSignature` should be called without contents first, because:

1. The underlying validation should never return a false valid, so the call is safe.
2. we cannot validate windows signatures, when called **WITH** contents.  So, we should try without contents first.
3. manipulating the contents makes it more likely to get a false invalid.

**Review by TravisEz13 — APPROVED** [review]

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 3]
```diff
@@ -516,15 +517,27 @@ private static void UntrustPublisher(Signature signature)
...
+
+            // try harder to validate the signature by being explicit about encoding
+            // and providing the script contents
+            string verificationContents = Encoding.Unicode.GetString(script.OriginalEncoding.GetPreamble()) + script.ScriptContents;
+            signature = SignatureHelper.GetSignature(path, verificationContents);
 
+            // A last ditch effort -
             // If the file was originally ASCII or UTF8, the SIP may have added the Unicode BOM
-            if ((signature.Status != SignatureStatus.Valid) && (script.OriginalEncoding != System.Text.Encoding.Unicode))
+            if ((signature.Status != SignatureStatus.Valid) && (script.OriginalEncoding != Encoding.Unicode))
```
Comment (rjmholt) [inline src/System.Management.Automation/security/SecurityManager.cs:538]:
```suggestion
            if (signature.Status != SignatureStatus.Valid
                && script.OriginalEncoding != Encoding.Unicode)
```

**Review by rjmholt — APPROVED** [review]

File: src/System.Management.Automation/security/SecurityManager.cs (hunk 2 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/security/SecurityManager.cs]: done

**Review by TravisEz13 — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@JamesWTruher Is there a way we can add a test for the change?

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 4]
```diff
@@ -516,15 +517,27 @@ private static void UntrustPublisher(Signature signature)
             }
         }
 
+        // Check the signature via the SIP which should never erroneously validate an invalid signature
+        // or altered script.
         private static Signature GetSignatureWithEncodingRetry(string path, ExternalScriptInfo script)
         {
-            string verificationContents = System.Text.Encoding.Unicode.GetString(script.OriginalEncoding.GetPreamble()) + script.ScriptContents;
-            Signature signature = SignatureHelper.GetSignature(path, verificationContents);
+            // Invoke the SIP directly with the most simple method
```
Comment (PaulHigin) [inline src/System.Management.Automation/security/SecurityManager.cs:524]: This comment is confusing to me, what is the simple method?  I think it should say something like 'first try to get a catalog signature, by passing in null content'.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Does this change need to be backported to Windows PowerShell?

**Comment by JamesWTruher** [comment]
@adityapatwardhan I couldn't think of one other than the tests we have, and I'm not sure why the tests we have didn't find this. It may be that @PaulHigin can figure out a way

File: src/System.Management.Automation/security/SecurityManager.cs (hunk 4 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/security/SecurityManager.cs:524]: My understanding is that the missing catalog look up is only part of the problem, and that we should be letting the SIP try to validate with out our "help". @TravisEz13 can perhaps address additional points.

**Comment by SteveL-MSFT** [comment]
@PaulHigin we initially checked against WinPS and for whatever reason it doesn't fail in WinPS.  I compared the code between WinPS and PS7 and don't see any differences in this specific code path so it remains a mystery why it only fails on PS7.

**Comment by PaulHigin** [comment]
@SteveL-MSFT  Hmm, I don't like those kinds of mysteries.  Could it be a PInvoke difference between .Net frameworks?  But this seems like the right fix since it matches what `Get-AuthenticodeSignature` does.

**Comment by rjmholt** [comment]
I think the difference lies here:

https://github.com/PowerShell/PowerShell/blob/3c7f4b573e96684781b8e1935eeb0b981bcd93bc/src/System.Management.Automation/engine/ExternalScriptInfo.cs#L518

**Comment by iSazonov** [comment]
I wonder that `Get-AuthenticodeSignature -Content` assumes the Content is always in Unicode.

> I think the difference lies here:
> 
> https://github.com/PowerShell/PowerShell/blob/3c7f4b573e96684781b8e1935eeb0b981bcd93bc/src/System.Management.Automation/engine/ExternalScriptInfo.cs#L518

It could be so if a test script file is in Unicode but without BOM otherwise StreamReader() detected right an encoding of the file.

**Comment by iSazonov** [comment]
Although it may be just the same
```powershell
[System.Text.Encoding] :: Utf8.GetPreamble ()
239
187
191
```
I mean, now we get the default BOM for Uft8 when it shouldn't be.

File: src/System.Management.Automation/security/SecurityManager.cs (hunk 4 above)
Comment (TravisEz13) [reply src/System.Management.Automation/security/SecurityManager.cs:524]: I think getting the catalog is only half of it, like James is saying.  If we fail to get a catalog signature, we try to get the embedded signature without modifying the content.  This should be what we try first.  It is the only thing we do in Get-AuthenticodeSignature.

**Comment by PaulHigin** [comment]
I see.  This does seem like it could be the culprit.  But the comment in `GetSignatureWithEncodingRetry` says the SIP will (may?) replace with a Unicode BOM when signing, in which case the code should handle it.  We should investigate this more.

**Comment by adityapatwardhan** [comment]
I am marking this PR as WIP as more discussion is required here.

**Review by PaulHigin — APPROVED** [review]
This change looks good to me.  I feel we can look into the SIP issue separately.

**Pull request merged**